### PR TITLE
C++: Improve Iterator models

### PR DESCRIPTION
**Pull request by geoffw0**
Improve Iterator.qll models:
 - for the most part `Iterator.qll` already models its classes, and uses a special predicate `getIteratorArgumentInput` to connect classes to non-member function models.  But there were a few opportunities to clean things up for member function models using `getClassAndName`.
 - there were some issues with the `operator+=` models, particularly with the non-member version and consistency between the member and non-member versions; I've added some more test cases and attempted to improve matters.  But please do say if you think I may have made a mistake!

Regarding BSL:
 - the `bsl` classes appear to have iterators, `begin` methods etc with the normal behaviour.
 - The `bsl::iterator`, `bsl::front_inserter`, `bsl::inserter`, and `bsl::back_inserter` functions don't appear to be documented but are [implemented as references](https://github.com/bloomberg/bde/blob/81e73ce7156bcffff46703ee08b478c774684423/groups/bsl/bslstl/bslstl_iterator.h#L195) (via `using` declarations) to the `std` versions.
 - depending on macros there is a [`bsl:iterator_traits`](https://github.com/bloomberg/bde/blob/81e73ce7156bcffff46703ee08b478c774684423/groups/bsl/bslstl/bslstl_iterator.h#L231) corresponding to `std::iterator_traits`, or a `using`.  I couldn't find documentation for it but it's clearly intended to be the same thing.
 - there are also a handful of special iterators (I think), such as `bsl::List_Iterator`.  This example should be caught by `IteratorByTypedefs` and modelled correctly.

TL;DR: `bsl` iterators work the same as `std` iterators, so I've changed the model to include them, and also fixed a few things.

### CODE REVIEW

**Review by criemen — APPROVED** [review]
LGTM, but somebody else should double-check the new taint flow.

**Comment by jbj** [comment]
> somebody else should double-check the new taint flow.

@rdmarsh2, can you check it?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll [hunk 1]
```diff
@@ -157,9 +157,11 @@ private class IteratorSubOperator extends Operator, TaintFunction {
  * A non-member `operator+=` or `operator-=` function for an iterator type.
  */
 private class IteratorAssignArithmeticOperator extends Operator, DataFlowFunction, TaintFunction {
+  FunctionInput iteratorInput;
+
   IteratorAssignArithmeticOperator() {
     this.hasName(["operator+=", "operator-="]) and
-    this.getDeclaringType() instanceof Iterator
+    iteratorInput = getIteratorArgumentInput(this, 0)
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:164]: The new field isn't referenced anywhere else - should it be deleted or did you plan to use it somewhere else?

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll]: Good point, I've removed it now.

**Review by criemen — APPROVED** [review]
As Robert didn't complain about the new taint flow, and his comment has been addressed, LGTM.
